jax/lib/xla_bridge.py

```diff
@@ -292,6 +292,41 @@ def constant(builder, py_val, canonicalize_types=True):
   else:
     raise TypeError("No constant handler for type: {}".format(py_type))
 
+# XLA supports arbitrary tuple nesting, but JAX only uses one level of tupling
+# (and our type checkers don't support recursive types).
+SpatialSharding = Union[Sequence[int], None, Tuple[Sequence[int], None]]
+
+def _sharding_to_proto(sharding: SpatialSharding):
+  """Converts a SpatialSharding to an OpSharding."""
+  proto = xla_client.OpSharding()
+  if isinstance(sharding, tuple):
+    if sharding[0] is None or isinstance(sharding[0], tuple):
+      sub_protos = [_sharding_to_proto(s) for s in sharding]
+      proto.type = xla_client.OpSharding.Type.TUPLE
+      proto.tuple_shardings = sub_protos
+      return proto
+
+  if sharding is None:
+    proto.type = xla_client.OpSharding.Type.REPLICATED
+  else:
+    proto.type = xla_client.OpSharding.Type.OTHER
+    proto.tile_assignment_dimensions = list(sharding)
+    proto.tile_assignment_devices = list(range(onp.product(sharding)))
+  return proto
+
+def set_sharding(builder, op, sharding: SpatialSharding):
+  """Uses CustomCall to annotate a value as sharded."""
```

Comment:
Is this an existing customcall? What does it do?

Reply:
It is. I added a comment.